File: 2463-minimum-total-distance-traveled/2463-minimum-total-distance-traveled.java
class Solution {
    //Dynamic Approach
    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
    Collections.sort(robot); 
    Arrays.sort(factory, Comparator.comparingInt(a -> a[0]));

    ArrayList<Integer> factoryPos = new ArrayList<>();
    for(int d[] : factory)
    for(int i=0; i<d[1]; i++)
    factoryPos.add(d[0]);
    
    int m = robot.size(), n = factoryPos.size();
    long dp[][] = new long[m+1][n+1];
   
    for(int i=0; i<m; i++)
    dp[i][n] = (long) 1e12;
    
    for(int i=m-1; i>=0; i--)
    for(int j=n-1; j>=0; j--)
    {
    long pick = Math.abs(robot.get(i)-factoryPos.get(j)) + dp[i+1][j+1];
    long notPick = dp[i][j+1];
    dp[i][j] = Math.min(pick, notPick);
    }
    return dp[0][0];
    }

//Recursion with memoization 
    public long memo(int robIndex, int facIndex, List<Integer> robot, ArrayList<Integer> factoryPos, long [][]dp)
    {
      if(robIndex>=robot.size())
      return 0;
      if(facIndex>=factoryPos.size())
      return (long)1e12;

      if(dp[robIndex][facIndex]!=-1)
      return dp[robIndex][facIndex];
    
      long pick = Math.abs(robot.get(robIndex)-factoryPos.get(facIndex)) + memo(robIndex+1, facIndex+1, robot, factoryPos, dp);

    long notPick = memo(robIndex, facIndex+1, robot, factoryPos, dp);
    return dp[robIndex][facIndex] = Math.min(pick, notPick);

    }
}

File: 2463-minimum-total-distance-traveled/2463-minimum-total-distance-traveled.cpp
#define ll long long
class Solution {
public:
    long long minimumTotalDistance(vector<int>& robot, vector<vector<int>>& factory, int idx = 0) {
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());
        vector<vector<vector<ll>>> dp(robot.size(), vector<vector<ll>>(factory.size(), vector<ll>(robot.size(), -1)));
        return fun(0, 0, 0, robot, factory, dp);
    }
    
    ll fun(int i, int j, int k, vector<int> &robot, vector<vector<int>> &factory, vector<vector<vector<ll>>> &dp) {
        if(i == robot.size()) return 0;
        if(j == factory.size()) return 1e16;
        if(dp[i][j][k] != -1) return dp[i][j][k];
        
        ll res1 = fun(i, j+1, 0, robot, factory, dp);
        ll res2 = (k < factory[j][1]) ? (abs(factory[j][0] - robot[i]) + fun(i+1, j, k+1, robot, factory, dp)) : 1e16;
        
        return dp[i][j][k] = min(res1, res2);
    }
};
